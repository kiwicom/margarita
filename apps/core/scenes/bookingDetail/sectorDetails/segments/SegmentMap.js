// @flow

import * as React from 'react';
import { MapView } from '@kiwicom/margarita-map';
import { StyleSheet } from '@kiwicom/universal-components';
import { graphql, createFragmentContainer } from '@kiwicom/margarita-relay';
import { getCenter } from 'geolib';
import memoize from 'memoize-one';
import { uniqBy, prop } from 'ramda';

import type { SegmentMap as BookingType } from './__generated__/SegmentMap.graphql';
import MapLines from './MapLines';
import SegmentMapMarker from './SegmentMapMarker';

type Props = {|
  +data: ?BookingType,
|};

type RouteStop = {|
  +stop: ?{|
    +city: ?{|
      +name: ?string,
    |},
    +locationId: ?string,
    +coordinates: ?{|
      +latitude: ?number,
      +longitude: ?number,
    |},
  |},
|};

type Segments = ?$ReadOnlyArray<?{|
  +arrival: ?RouteStop,
  +departure: ?RouteStop,
|}>;

type MarkerData = {|
  +latitude: ?number,
  +longitude: ?number,
  +name: ?string,
  +locationId: string,
|};

type CompareByType = {|
  +locationId: string,
|};

const DEFAULT_PADDING = { top: 40, right: 40, bottom: 40, left: 40 };

const mapMarkerData = memoize((props: Props) => {
  switch (props.data?.type) {
    case 'BOOKING_ONE_WAY':
      return mapSegmentsToMarkerData(props.data?.sector?.segments);
    case 'BOOKING_RETURN':
      return mapSegmentsToMarkerData(getReturnSegments(props));
    case 'BOOKING_MULTICITY':
      return mapSegmentsToMarkerData(getMulticitySegments(props));
    default:
      return null;
  }
});

const getReturnSegments = (props: Props) => {
  const outbound = props.data?.outbound?.segments ?? [];
  const inbound = props.data?.inbound?.segments ?? [];
  return [...outbound, ...inbound];
};

const getMulticitySegments = (props: Props) => {
  const sectors = props.data?.sectors ?? [];
  return sectors.reduce((acc, curr) => {
    const segments = curr?.segments ?? [];
    return [...acc, ...segments];
  }, []);
};

const mapSegmentsToMarkerData = (segments: Segments) => {
  if (segments == null) {
    return [];
  }
  const bookingSegments = segments.reduce((acc, curr) => {
    return [
      ...acc,
      {
        ...curr?.departure?.stop?.coordinates,
        name: curr?.departure?.stop?.city?.name,
        locationId: curr?.departure?.stop?.locationId ?? '',
      },
      {
        ...curr?.arrival?.stop?.coordinates,
        name: curr?.arrival?.stop?.city?.name,
        locationId: curr?.arrival?.stop?.locationId ?? '',
      },
    ];
  }, []);

  return uniqBy<MarkerData, CompareByType>(
    prop<string, MarkerData>('locationId'),
    bookingSegments,
  );
};

export class SegmentMap extends React.Component<Props> {
  map: ?React.Ref<typeof MapView>;

  setRef = (ref: ?React.Ref<typeof MapView>) => {
    this.map = ref;
  };

  getMarkerData = () => mapMarkerData(this.props);

  fit = () => {
    if (this.map != null) {
      // $FlowExpectedError: fitToCoordinates does exist on map, but flow is confused by .native|.web exports
      this.map.fitToCoordinates(this.getMarkerData(), {
        edgePadding: DEFAULT_PADDING,
      });
    }
  };

  render() {
    const locations = this.getMarkerData() ?? [];
    const center = getCenter(this.getMarkerData());
    return (
      <MapView
        ref={this.setRef}
        style={styles.mapView}
        initialRegion={{
          latitude: parseFloat(center.latitude),
          longitude: parseFloat(center.longitude),
          latitudeDelta: 1,
          longitudeDelta: 1,
        }}
        zoomEnabled={false}
        rotateEnabled={false}
        pitchEnabled={false}
        onMapReady={this.fit}
      >
        {locations.map(location => (
          <SegmentMapMarker
            key={location?.locationId}
            latitude={location?.latitude}
            longitude={location?.longitude}
            name={location?.name}
          />
        ))}
        <MapLines data={this.props.data} />
      </MapView>
    );
  }
}

const styles = StyleSheet.create({
  mapView: {
    flex: 1,
    height: 200,
  },
});

export default createFragmentContainer(
  SegmentMap,
  graphql`
    fragment SegmentMapStop on RouteStop {
      stop {
        locationId
        city {
          name
        }
        coordinates {
          latitude: lat
          longitude: lng
        }
      }
    }
    fragment SegmentMapSegment on Sector {
      segments {
        arrival {
          ...SegmentMapStop @relay(mask: false)
        }
        departure {
          ...SegmentMapStop @relay(mask: false)
        }
      }
    }
    fragment SegmentMap on BookingInterface {
      ...MapLines
      type
      ... on BookingOneWay {
        sector {
          ...SegmentMapSegment @relay(mask: false)
        }
      }
      ... on BookingReturn {
        inbound {
          ...SegmentMapSegment @relay(mask: false)
        }
        outbound {
          ...SegmentMapSegment @relay(mask: false)
        }
      }
      ... on BookingMulticity {
        sectors {
          ...SegmentMapSegment @relay(mask: false)
        }
      }
    }
  `,
);
