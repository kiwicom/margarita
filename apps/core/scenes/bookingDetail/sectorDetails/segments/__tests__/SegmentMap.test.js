// @flow

import * as React from 'react';
import { shallow } from 'react-native-testing-library';

import { SegmentMap } from '../SegmentMap';

type BookingType = 'BOOKING_MULTICITY' | 'BOOKING_ONE_WAY' | 'BOOKING_RETURN';

const $fragmentRefs: any = null;
const $refType: any = null;

const getArrival = (index: number) => ({
  stop: {
    locationId: `arrival-${index}`,
    city: {
      name: `arrivalCity-${index}`,
    },
    coordinates: {
      latitude: 1 + index,
      longitude: 2 + index,
    },
  },
});
const getDeparture = (index: number) => ({
  stop: {
    locationId: `departure-${index}`,
    city: {
      name: `departure-${index}`,
    },
    coordinates: {
      latitude: 10 + index,
      longitude: 20 + index,
    },
  },
});
const getData = (type: BookingType) => ({
  $fragmentRefs,
  $refType,
  type,
  sector: {
    segments: [
      {
        arrival: getArrival(1),
        departure: getDeparture(1),
      },
    ],
  },
  inbound: {
    segments: [{ arrival: getArrival(2), departure: getDeparture(2) }],
  },
  outbound: {
    segments: [{ arrival: getArrival(3), departure: getDeparture(3) }],
  },
  sectors: [
    {
      segments: [
        {
          arrival: getArrival(4),
          departure: getDeparture(4),
        },
      ],
    },
    {
      segments: [
        {
          arrival: getArrival(5),
          departure: getDeparture(5),
        },
      ],
    },
  ],
});

it('renders correctly with OneWayBooking', () => {
  const data = getData('BOOKING_ONE_WAY');
  const { output } = shallow(<SegmentMap data={data} />);
  expect(output).toMatchSnapshot();
});

it('renders correctly with ReturnBooking', () => {
  const data = getData('BOOKING_RETURN');
  const { output } = shallow(<SegmentMap data={data} />);
  expect(output).toMatchSnapshot();
});

it('renders correctly with MulticityBooking', () => {
  const data = getData('BOOKING_MULTICITY');
  const { output } = shallow(<SegmentMap data={data} />);
  expect(output).toMatchSnapshot();
});
