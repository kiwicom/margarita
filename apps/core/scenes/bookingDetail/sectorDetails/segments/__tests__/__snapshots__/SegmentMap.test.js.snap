// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with MulticityBooking 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 10.122433,
      "latitudeDelta": 1,
      "longitude": 15.373298,
      "longitudeDelta": 1,
    }
  }
  onMapReady={[Function]}
  pitchEnabled={false}
  rotateEnabled={false}
  style={
    Object {
      "flex": 1,
      "height": 200,
    }
  }
  zoomEnabled={false}
>
  <SegmentMapMarker
    latitude={14}
    longitude={24}
    name="departure-4"
  />
  <SegmentMapMarker
    latitude={5}
    longitude={6}
    name="arrivalCity-4"
  />
  <SegmentMapMarker
    latitude={15}
    longitude={25}
    name="departure-5"
  />
  <SegmentMapMarker
    latitude={6}
    longitude={7}
    name="arrivalCity-5"
  />
  <ForwardRef(Relay(MapLines))
    data={
      Object {
        "$fragmentRefs": null,
        "$refType": null,
        "inbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-2",
                  },
                  "coordinates": Object {
                    "latitude": 3,
                    "longitude": 4,
                  },
                  "locationId": "arrival-2",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-2",
                  },
                  "coordinates": Object {
                    "latitude": 12,
                    "longitude": 22,
                  },
                  "locationId": "departure-2",
                },
              },
            },
          ],
        },
        "outbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-3",
                  },
                  "coordinates": Object {
                    "latitude": 4,
                    "longitude": 5,
                  },
                  "locationId": "arrival-3",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-3",
                  },
                  "coordinates": Object {
                    "latitude": 13,
                    "longitude": 23,
                  },
                  "locationId": "departure-3",
                },
              },
            },
          ],
        },
        "sector": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-1",
                  },
                  "coordinates": Object {
                    "latitude": 2,
                    "longitude": 3,
                  },
                  "locationId": "arrival-1",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-1",
                  },
                  "coordinates": Object {
                    "latitude": 11,
                    "longitude": 21,
                  },
                  "locationId": "departure-1",
                },
              },
            },
          ],
        },
        "sectors": Array [
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-4",
                    },
                    "coordinates": Object {
                      "latitude": 5,
                      "longitude": 6,
                    },
                    "locationId": "arrival-4",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-4",
                    },
                    "coordinates": Object {
                      "latitude": 14,
                      "longitude": 24,
                    },
                    "locationId": "departure-4",
                  },
                },
              },
            ],
          },
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-5",
                    },
                    "coordinates": Object {
                      "latitude": 6,
                      "longitude": 7,
                    },
                    "locationId": "arrival-5",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-5",
                    },
                    "coordinates": Object {
                      "latitude": 15,
                      "longitude": 25,
                    },
                    "locationId": "departure-5",
                  },
                },
              },
            ],
          },
        ],
        "type": "BOOKING_MULTICITY",
      }
    }
  />
</MapView>
`;

exports[`renders correctly with OneWayBooking 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 6.58031,
      "latitudeDelta": 1,
      "longitude": 11.918627,
      "longitudeDelta": 1,
    }
  }
  onMapReady={[Function]}
  pitchEnabled={false}
  rotateEnabled={false}
  style={
    Object {
      "flex": 1,
      "height": 200,
    }
  }
  zoomEnabled={false}
>
  <SegmentMapMarker
    latitude={11}
    longitude={21}
    name="departure-1"
  />
  <SegmentMapMarker
    latitude={2}
    longitude={3}
    name="arrivalCity-1"
  />
  <ForwardRef(Relay(MapLines))
    data={
      Object {
        "$fragmentRefs": null,
        "$refType": null,
        "inbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-2",
                  },
                  "coordinates": Object {
                    "latitude": 3,
                    "longitude": 4,
                  },
                  "locationId": "arrival-2",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-2",
                  },
                  "coordinates": Object {
                    "latitude": 12,
                    "longitude": 22,
                  },
                  "locationId": "departure-2",
                },
              },
            },
          ],
        },
        "outbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-3",
                  },
                  "coordinates": Object {
                    "latitude": 4,
                    "longitude": 5,
                  },
                  "locationId": "arrival-3",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-3",
                  },
                  "coordinates": Object {
                    "latitude": 13,
                    "longitude": 23,
                  },
                  "locationId": "departure-3",
                },
              },
            },
          ],
        },
        "sector": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-1",
                  },
                  "coordinates": Object {
                    "latitude": 2,
                    "longitude": 3,
                  },
                  "locationId": "arrival-1",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-1",
                  },
                  "coordinates": Object {
                    "latitude": 11,
                    "longitude": 21,
                  },
                  "locationId": "departure-1",
                },
              },
            },
          ],
        },
        "sectors": Array [
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-4",
                    },
                    "coordinates": Object {
                      "latitude": 5,
                      "longitude": 6,
                    },
                    "locationId": "arrival-4",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-4",
                    },
                    "coordinates": Object {
                      "latitude": 14,
                      "longitude": 24,
                    },
                    "locationId": "departure-4",
                  },
                },
              },
            ],
          },
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-5",
                    },
                    "coordinates": Object {
                      "latitude": 6,
                      "longitude": 7,
                    },
                    "locationId": "arrival-5",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-5",
                    },
                    "coordinates": Object {
                      "latitude": 15,
                      "longitude": 25,
                    },
                    "locationId": "departure-5",
                  },
                },
              },
            ],
          },
        ],
        "type": "BOOKING_ONE_WAY",
      }
    }
  />
</MapView>
`;

exports[`renders correctly with ReturnBooking 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 8.09869,
      "latitudeDelta": 1,
      "longitude": 13.399013,
      "longitudeDelta": 1,
    }
  }
  onMapReady={[Function]}
  pitchEnabled={false}
  rotateEnabled={false}
  style={
    Object {
      "flex": 1,
      "height": 200,
    }
  }
  zoomEnabled={false}
>
  <SegmentMapMarker
    latitude={13}
    longitude={23}
    name="departure-3"
  />
  <SegmentMapMarker
    latitude={4}
    longitude={5}
    name="arrivalCity-3"
  />
  <SegmentMapMarker
    latitude={12}
    longitude={22}
    name="departure-2"
  />
  <SegmentMapMarker
    latitude={3}
    longitude={4}
    name="arrivalCity-2"
  />
  <ForwardRef(Relay(MapLines))
    data={
      Object {
        "$fragmentRefs": null,
        "$refType": null,
        "inbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-2",
                  },
                  "coordinates": Object {
                    "latitude": 3,
                    "longitude": 4,
                  },
                  "locationId": "arrival-2",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-2",
                  },
                  "coordinates": Object {
                    "latitude": 12,
                    "longitude": 22,
                  },
                  "locationId": "departure-2",
                },
              },
            },
          ],
        },
        "outbound": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-3",
                  },
                  "coordinates": Object {
                    "latitude": 4,
                    "longitude": 5,
                  },
                  "locationId": "arrival-3",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-3",
                  },
                  "coordinates": Object {
                    "latitude": 13,
                    "longitude": 23,
                  },
                  "locationId": "departure-3",
                },
              },
            },
          ],
        },
        "sector": Object {
          "segments": Array [
            Object {
              "arrival": Object {
                "stop": Object {
                  "city": Object {
                    "name": "arrivalCity-1",
                  },
                  "coordinates": Object {
                    "latitude": 2,
                    "longitude": 3,
                  },
                  "locationId": "arrival-1",
                },
              },
              "departure": Object {
                "stop": Object {
                  "city": Object {
                    "name": "departure-1",
                  },
                  "coordinates": Object {
                    "latitude": 11,
                    "longitude": 21,
                  },
                  "locationId": "departure-1",
                },
              },
            },
          ],
        },
        "sectors": Array [
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-4",
                    },
                    "coordinates": Object {
                      "latitude": 5,
                      "longitude": 6,
                    },
                    "locationId": "arrival-4",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-4",
                    },
                    "coordinates": Object {
                      "latitude": 14,
                      "longitude": 24,
                    },
                    "locationId": "departure-4",
                  },
                },
              },
            ],
          },
          Object {
            "segments": Array [
              Object {
                "arrival": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "arrivalCity-5",
                    },
                    "coordinates": Object {
                      "latitude": 6,
                      "longitude": 7,
                    },
                    "locationId": "arrival-5",
                  },
                },
                "departure": Object {
                  "stop": Object {
                    "city": Object {
                      "name": "departure-5",
                    },
                    "coordinates": Object {
                      "latitude": 15,
                      "longitude": 25,
                    },
                    "locationId": "departure-5",
                  },
                },
              },
            ],
          },
        ],
        "type": "BOOKING_RETURN",
      }
    }
  />
</MapView>
`;
